### ISSUE TITLE
蒸馏效果不好，请问一下怎么解决？

### ISSUE THREAD

**Issue by cgq0816** [airaria/TextBrewer#33]
1、使用官方提供参数
蒸馏用的学习率 lr=1e-4(除非特殊说明)。训练30~60轮。
2、语料集大小 1W左右标注数据
3、模型加载
        # Define models
        bert_config = BertConfig.from_json_file('bert/bert_config/bert_config.json')
        bert_config_T6 = BertConfig.from_json_file('bert/bert_config/bert_config_T6.json')
        
        bert_config.output_hidden_states = True
        bert_config_T6.output_hidden_states = True  
        
        bert_config.num_labels=self.num_labels
        bert_config_T6.num_labels=self.num_labels
        
        teacher_model = BertForSequenceClassification.from_pretrained(self.distill_bert_path,config=bert_config)
        student_model = BertForSequenceClassification(bert_config_T6) #, num_labels = 2

4、实验结果
学生模型最终在数据集上的准确率 0.14，如果用大模型fine-turning的话也是一个很好的效果，不知道蒸馏之后的效果为什么那么差，还是我使用的方式不对

蒸馏代码
`    def distill_fit(self, train_df,dev_df,*args):
        self.get_model_config()
        
        if self.set_en_train:
            if not self.tokenizer:self.tokenizer = BertTokenizer.from_pretrained(self.en_vocab_path)
        else:
            if not self.tokenizer:self.tokenizer = BertTokenizer.from_pretrained(self.distill_bert_path)
            
        #检测gpu是否可用
        device=PlatformUtils.get_device()
        logging.info('--------cuda device--------:%s'%(device))
        device='cuda:0'
        
        self.device = torch.device(device if device else "cpu")
        
        train_it,dev_it,train_dl=self.transform(train_df,dev_df,*args)
        
        
        # Define models
        bert_config = BertConfig.from_json_file('bert/bert_config/bert_config.json')
        bert_config_T6 = BertConfig.from_json_file('bert/bert_config/bert_config_T6.json')
        
        bert_config.output_hidden_states = True
        bert_config_T6.output_hidden_states = True  
        
        bert_config.num_labels=self.num_labels
        bert_config_T6.num_labels=self.num_labels
        
        #teacher_model = BertForSequenceClassification(bert_config) #, num_labels = 2 self.bert_path
        teacher_model = BertForSequenceClassification.from_pretrained(self.distill_bert_path,config=bert_config)
        # Teacher should be initialized with pre-trained weights and fine-tuned on the downstream task.
        # For the demonstration purpose, we omit these steps here
        
        student_model = BertForSequenceClassification(bert_config_T6) #, num_labels = 2
        
        teacher_model.to(device=self.device)
        student_model.to(device=self.device)
        
        # Optimizer and learning rate scheduler
        optimizer = AdamW(student_model.parameters(), lr=1e-4)
        
        
        if not self.criterion:self.criterion=nn.CrossEntropyLoss()
        
        num_epochs = 60
        num_training_steps = len(train_it) * num_epochs
        
        scheduler_class = get_linear_schedule_with_warmup
        # arguments dict except 'optimizer'
        scheduler_args = {'num_warmup_steps':int(0.1*num_training_steps), 'num_training_steps':num_training_steps}
        
        # display model parameters statistics
        print("\nteacher_model's parametrers:")
        result, _ = textbrewer.utils.display_parameters(teacher_model,max_level=3)
        print (result)
        print("student_model's parametrers:")
        result, _ = textbrewer.utils.display_parameters(student_model,max_level=3)
        print (result)
        from functools import partial
        callback_fun = partial(self.validate, eval_dl=train_dl) # fill other arguments
        # Initialize configurations and distiller
        train_config = TrainingConfig(device=self.device)
        distill_config = DistillationConfig(
            temperature=8,
            hard_label_weight=0,
            kd_loss_type='ce',
            probability_shift=False,
            is_caching_logits=True,
            intermediate_matches=[
                {"layer_T":0, "layer_S":0, "feature":"hidden", "loss":"hidden_mse", "weight":1}, 
               {"layer_T":2, "layer_S":1, "feature":"hidden", "loss":"hidden_mse", "weight":1}, 
               {"layer_T":4, "layer_S":2, "feature":"hidden", "loss":"hidden_mse", "weight":1}, 
               {"layer_T":6, "layer_S":3, "feature":"hidden", "loss":"hidden_mse", "weight":1}, 
               {"layer_T":8, "layer_S":4, "feature":"hidden", "loss":"hidden_mse", "weight":1}, 
               {"layer_T":10,"layer_S":5, "feature":"hidden", "loss":"hidden_mse", "weight":1}, 
               {"layer_T":12,"layer_S":6, "feature":"hidden", "loss":"hidden_mse", "weight":1}]
        )
        
        print ("train_config:")
        print (train_config)
        
        print ("distill_config:")
        print (distill_config)
        
        distiller = GeneralDistiller(
            train_config=train_config, distill_config = distill_config,
            model_T = teacher_model, model_S = student_model, 
            adaptor_T = self.simple_adaptor, adaptor_S = self.simple_adaptor)
        
        # Start distilling
        with distiller:
            distiller.train(optimizer,train_it, num_epochs=num_epochs, 
            scheduler_class=scheduler_class, scheduler_args = scheduler_args, callback=callback_fun) `

**Comment by airaria** [comment]
从这段代码看没什么问题。
几个可能出错的地方在于：
1. Student模型这里是随机初始化的。尝试用BERT前6层初始化，看看效果有没有什么变化？
2. 用最简单的BasicDistiller实验：student用BERT前6层初始化，只用kd_loss（不加intermediate_matches）
3. 可以展示下adaptor，它的返回和模型的输出是否一致？

 最后，可以参考例子中MNLI任务的蒸馏。

**Comment by cgq0816** [comment]
使用random_tokens_example做了几组实验
1、蒸馏到6层，不使用intermediate_matches
2、蒸馏到3层，使用intermediate_matches（默认random_tokens_example蒸馏方式）
3、蒸馏到6层，使用intermediate_matches（使用自己配置蒸馏方式）

以下蒸馏的结果
1、蒸馏到6层，不使用intermediate_matches
distill_config:
temperature : 8
temperature_scheduler : None
hard_label_weight : 0
hard_label_weight_scheduler : None
kd_loss_type : ce
kd_loss_weight : 1
kd_loss_weight_scheduler : None
probability_shift : False
intermediate_matches : []
is_caching_logits : False
第一次蒸馏结果：
Number of examples:  100
Acc:  0.56
第二次蒸馏结果：
Number of examples:  100
Acc:  0.45
两次蒸馏的结果随机性比较大
2、蒸馏到3层，使用intermediate_matches
distill_config:
temperature : 8
temperature_scheduler : None
hard_label_weight : 0
hard_label_weight_scheduler : None
kd_loss_type : ce
kd_loss_weight : 1
kd_loss_weight_scheduler : None
probability_shift : False
intermediate_matches : [
IntermediateMatch: layer_T : 0, layer_S : 0, feature : hidden, weight : 1, loss : hidden_mse, proj : None, 
IntermediateMatch: layer_T : 8, layer_S : 2, feature : hidden, weight : 1, loss : hidden_mse, proj : None, 
IntermediateMatch: layer_T : [0, 0], layer_S : [0, 0], feature : hidden, weight : 1, loss : nst, proj : None, 
IntermediateMatch: layer_T : [8, 8], layer_S : [2, 2], feature : hidden, weight : 1, loss : nst, proj : None]
is_caching_logits : False

第一次蒸馏结果：
Number of examples:  100
Acc:  0.53
3、蒸馏到6层，使用intermediate_matches
distill_config:
temperature : 8
temperature_scheduler : None
hard_label_weight : 0
hard_label_weight_scheduler : None
kd_loss_type : ce
kd_loss_weight : 1
kd_loss_weight_scheduler : None
probability_shift : False
intermediate_matches : [
IntermediateMatch: layer_T : 0, layer_S : 0, feature : hidden, weight : 1, loss : hidden_mse, proj : None, 
IntermediateMatch: layer_T : 2, layer_S : 1, feature : hidden, weight : 1, loss : hidden_mse, proj : None, 
IntermediateMatch: layer_T : 4, layer_S : 2, feature : hidden, weight : 1, loss : hidden_mse, proj : None, 
IntermediateMatch: layer_T : 6, layer_S : 3, feature : hidden, weight : 1, loss : hidden_mse, proj : None, 
IntermediateMatch: layer_T : 8, layer_S : 4, feature : hidden, weight : 1, loss : hidden_mse, proj : None, 
IntermediateMatch: layer_T : 10, layer_S : 5, feature : hidden, weight : 1, loss : hidden_mse, proj : None, 
IntermediateMatch: layer_T : 12, layer_S : 6, feature : hidden, weight : 1, loss : hidden_mse, proj : None]
is_caching_logits : False

第一次蒸馏结果：
Number of examples:  100
Acc:  0.6

我使用的adaptor
    def simple_adaptor(self,batch, model_outputs):
        # The second element of model_outputs is the logits before softmax
        # The third element of model_outputs is hidden states
        return {'logits': model_outputs[1],
                'hidden': model_outputs[2],
                'inputs_mask': batch['attention_mask']}   
和random_tokens_example使用是一样的

**Comment by airaria** [comment]
方便的话你可以把脱敏代码发我邮箱我有空查一查

**Comment by cgq0816** [comment]
麻烦了，里面包含数据集和代码，不过代码需要微调，实际不能运行，缺少相关依赖文件，谢谢帮助


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "airaria/TextBrewer"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年12月30日(星期三) 晚上6:22
收件人:&nbsp;"airaria/TextBrewer"<TextBrewer@noreply.github.com&gt;;
抄送:&nbsp;"云"<276851182@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [airaria/TextBrewer] 蒸馏效果不好，请问一下怎么解决？ (#33)





 
方便的话你可以把脱敏代码发我邮箱我有空查一查

**Comment by cgq0816** [comment]
不好意思，代码已调整，可以直接运行，麻烦在这个版本上看一下，谢谢啦


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "airaria/TextBrewer"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年12月30日(星期三) 晚上6:22
收件人:&nbsp;"airaria/TextBrewer"<TextBrewer@noreply.github.com&gt;;
抄送:&nbsp;"云"<276851182@qq.com&gt;;"Author"<author@noreply.github.com&gt;;
主题:&nbsp;Re: [airaria/TextBrewer] 蒸馏效果不好，请问一下怎么解决？ (#33)





 
方便的话你可以把脱敏代码发我邮箱我有空查一查
